Extension/src/LanguageServer/referencesView.ts

```diff
@@ -22,51 +23,59 @@ export class FindAllRefsView {
-
-    toggleGroupView(): void {
-        this.referenceViewProvider.toggleGroupView();
+        let hasResults: boolean = false;
+        if (this.referencesModel) {
+            hasResults = this.referencesModel.hasResults();
+        }
+        vscode.commands.executeCommand('setContext', 'cppReferenceTypes:hasResults', hasResults);
     }
 
-    setData(results: ReferencesResult, isCanceled: boolean): void {
-        this.referenceViewProvider.setModel(results, isCanceled);
+    setData(results: ReferencesResult, isCanceled: boolean, groupByFile: boolean): void {
+        this.referencesModel = new ReferencesModel(results, false, isCanceled, groupByFile, null, () => { this.referenceViewProvider.refresh(); });
+        this.referenceViewProvider.setModel(this.referencesModel);
     }
 
     clearData(): void {
         this.referenceViewProvider.clear();
     }
 
+    setGroupBy(groupByFile: boolean): void {
+        if (this.referencesModel) {
+            this.referencesModel.groupByFile = groupByFile;
+            this.referenceViewProvider.refresh();
+        }
+    }
+
     getResultsAsText(includeConfirmedReferences: boolean): string {
         let results: string[] = [];
         let confirmedRefs: string[] = [];
         let otherRefs: string[] = [];
         let fileRefs: string[] = [];
 
-        let referenceItems: ReferenceItem[] = this.referenceViewProvider.getReferenceItems();
-        for (let ref of referenceItems) {
-            let line: string =
-                ("[" + getReferenceTagString(ref.type, this.referenceViewProvider.isCanceled()) + "] "
-                + ref.parent.name
-                + ":" + (ref.position.line + 1) + ":" + (ref.position.character + 1)
-                + " " + ref.text);
-            if (includeConfirmedReferences && ref.type === ReferenceType.Confirmed) {
-                confirmedRefs.push(line);
-            } else {
-                otherRefs.push(line);
+        for (let ref of this.referencesModel.nodes) {
+            if (ref.node === NodeType.reference) {
+                let line: string =
+                    ("[" + getReferenceTagString(ref.referenceType, this.referencesModel.isCanceled) + "] "
+                    + ref.filename
+                    + ":" + (ref.referencePosition.line + 1) + ":" + (ref.referencePosition.character + 1)
+                    + " " + ref.referenceText);
+                if (includeConfirmedReferences && ref.referenceType === ReferenceType.Confirmed) {
+                    confirmedRefs.push(line);
+                } else {
+                    otherRefs.push(line);
+                }
             }
         }
 
         // Get files with pending references items (location of reference is pending)
-        let fileReferences: FileItem[] = this.referenceViewProvider.getFilesWithPendingReferences();
+        let fileReferences: TreeNode[] = this.referencesModel.nodes.filter(i => (i.referencePosition.line === 0 && i.referencePosition.character === 0));

```

Comment:
>this.referencesModel.nodes.filter(i => (i.referencePosition.line === 0 && i.referencePosition.character === 0)); [](start = 41, length = 112)

make nodes object private and create function getFileReferences() for consistency with other TreeNode item getters